src/table/hooks/use-snapshot.ts

```diff
@@ -12,37 +13,54 @@ interface UseSnapshotProps {
   contentRect: stardust.Rect;
 }
 
-const useSnapshot = ({ layout, viewService, model, rootElement, contentRect }: UseSnapshotProps) => {
-  const getViewState = () => {
-    if (viewService.viewState) return viewService.viewState;
-    if (layout.usePagination) {
-      const totalsPosition = getTotalPosition(layout);
-      const { visibleRowStartIndex = -1, visibleRowEndIndex = -1 } = findPaginationVisibleRows(
-        rootElement,
-        totalsPosition
-      );
-      return {
-        scrollLeft: viewService.scrollLeft,
-        visibleTop: viewService.qTop + visibleRowStartIndex,
-        visibleHeight: visibleRowEndIndex < 0 ? 0 : visibleRowEndIndex - visibleRowStartIndex + 1,
-        rowsPerPage: viewService.rowsPerPage,
-        page: viewService.page,
-      };
-    }
-    const { visibleRowStartIndex = -1, visibleRowEndIndex = -1 } = findVirtualizedVisibleRows(rootElement, viewService);
+const EXTRA_ROWS = 3;
+
+export const getVisibleHeight = (
+  visibleRowEndIndex: number,
+  visibleRowStartIndex: number,
+  layout: TableLayout,
+  viewService: ViewService
+) => {
+  if (visibleRowEndIndex < 0) return 0;
+
+  const totalRowCount = layout.qHyperCube.qSize.qcy;
+  const visualRowsPerPage = viewService.rowsPerPage || initialPageInfo.rowsPerPage;
+  // EXTRA_ROWS will be added to the visualHeight when the pagination footer is displayed and rows per page is heighr
+  return Math.min(totalRowCount, visualRowsPerPage, visibleRowEndIndex - visibleRowStartIndex + 1 + EXTRA_ROWS);
+};
+
+export const getViewState = (layout: TableLayout, viewService: ViewService, rootElement: HTMLElement) => {
+  if (viewService.viewState) return viewService.viewState;
```

Comment:
Do we have the case that `viewState` does not exist ? If so we should set the default values instead.
The condition doesn't sound proper! you should define `else` to handle the undefined case otherwise` it won't make sense. It could be something like this:
```suggestion
  viewService.viewState ?? { scrollLeft: 0, visibleLeft: 0 .... and others with their default values};
```

Reply:
I'm actually thinking the function is there to define the `viewState` as it returns it's properties. so I think if you want to have such a sanity check you should start it with `if(!vieService.viewState) { the rest of the code }` otherwise take the default values.